lib/src/common/cache/cache.dart

```diff
@@ -294,6 +376,47 @@ class Cache<TIdentifier, TValue> extends Object with Disposable {
     return new Future.value();
   }
 
+  /// Returns `true` and calls [callback] if there is a cached, unreleased
+  /// [TValue] associated with [id]; returns `false` otherwise.
+  ///
+  /// This does not perform a [get] or [getAsync], and as a result, will not
+  /// affect retention or removal of a [TIdentifier][TValue] pair from the cache.
+  ///
+  /// Any [TIdentifier] [TValue] pair removals will wait for the Future returned by
+  /// the call to [callback] before emitting [didRemove] events.
+  ///
+  /// If the [Cache] [isOrWillBeDisposed] then a [StateError] is thrown.
+  Future<bool> applyToItem(
+      TIdentifier id, dynamic callback(Future<TValue> value)) {
+    _log.finest('applyToItem id: $id');
+    _throwWhenDisposed('applyToItem');
+    if (_isReleased[id] != false) {
+      return new Future<bool>.value(false);
+    }
+
+    final callBackResult = callback(_cache[id]);
+    if (callBackResult is Future) {
+      // In this case we're only interested in the computation being done or not
+      // done, not in the result
+      final errorlessCallbackResult = callBackResult.catchError((_) {});
```

Comment:
Is there ever a scenario where you would want to log an error thrown by the callback?

Reply:
If so, it might be a good idea to use `.whenComplete` to allow this to propagate as an uncaught async error, which can be picked up by a `runZoned`'s `onError` handler and logged.

Reply:
Yup, I totally agree that the cache shouldn't care if the callback errors. The problem with the current code is that calling `.catchError` prevents that error from propagating to a zone handler.

I think that we should use `whenComplete` instead, which has the same behavior as the cache currently has, while allowing the error to propagate to consumer zones instead of being swallowed by `catchError`.
```dart
final completer = new Completer();
final errorlessCallbackResult = completer.future;
callBackResult.whenComplete(completer.complete);
```

---

Illustration of this behavior (which I wrote to make sure I understood it correctly 😄):

```dart
import 'dart:async';

main() {
  runZoned(() {
    handleFutureWhenComplete(new Future.error('error 1'));
    handleFutureCatchError(new Future.error('error 2'));
  }, onError: (error) {
    print('Zone onError: $error');
  });  
}

void handleFutureWhenComplete(Future future) {
  var completer = new Completer();
  var errorlessCallbackFuture = completer.future;
  
  future.whenComplete(completer.complete);
  
  errorlessCallbackFuture.then((_) {
    print('errorlessCallbackFuture (whenComplete) completed');
  }); 
}

void handleFutureCatchError(Future future) {
  var errorlessCallbackFuture = future.catchError((_) {});  
  errorlessCallbackFuture.then((_) {
    print('errorlessCallbackFuture (catchError) completed');
  }); 
}
```
Prints:
```
Zone onError: error 1
errorlessCallbackFuture (catchError) completed
errorlessCallbackFuture (whenComplete) completed
```

Reply:
Ah yeah, but we're returning the original future to the callers zone on line [415](https://github.com/Workiva/w_common/pull/90/files/971bf535b30057780368fe9d8322bd6825665334#diff-1fdd0bfc9ee195804c63d431b8409625R415) not the new future we created with `catchError`.

So what we have is more like this: 

```dart
import 'dart:async';

main() {
  runZoned(() {
    handleFutureWhenComplete(new Future.error('error 1'));
    handleFutureCatchError(new Future.error('error 2'));
  }, onError: (error) {
    print('Zone onError: $error');
  });  
}

void handleFutureWhenComplete(Future future) {
  var completer = new Completer();
  var errorlessCallbackFuture = completer.future;
  
  future.whenComplete(completer.complete);
  
  errorlessCallbackFuture.then((_) {
    print('errorlessCallbackFuture (whenComplete) completed');
  }); 
}

void handleFutureCatchError(Future future) {
  var errorlessCallbackFuture = future.catchError((_) {});
  errorlessCallbackFuture.then((_){
     print('errorlessCallbackFuture (catchError) completed');
  });
  future.then((_) {
    print('future completed');
  }); 
}
```
Prints
```
Zone onError: error 1
errorlessCallbackFuture (catchError) completed
Zone onError: error 2
errorlessCallbackFuture (whenComplete) completed
```

Reply:
That being said, maybe the `handleFutureWhenComplete` version is more idiomatic ¯\\\_(ツ)_/¯